### PR TITLE
Creates the Steering Wheel Widget

### PR DESCRIPTION
**Pull request by Voldivh**
## Summary
This PR addresses part of the issue https://github.com/ekumenlabs/gauges2/issues/1. As the tittle says, creates the widget for the steering wheel.

### CODE REVIEW

File: src/rqt_gauges_2/steering_wheel.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+    def __init__(self, context):
+        super(SteeringWheel, self).__init__(context)
+        self.setObjectName('Speedometer')
+
+        self._context = context
+        self._node = context.node
+
+        self._widget = SteeringWheelWidget(self._node)
+        context.add_widget(self._widget)
```
Comment (agalbachicar) [inline src/rqt_gauges_2/steering_wheel.py]: Nit: new blank line at the end of the file.

File: src/rqt_gauges_2/steering_wheel_gauge.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+        self.angle_offset = 0
+
+        self.minValue = -45
+        self.maxValue = 45
+
+        if self.width <= self.height:
+            self.widget_diameter = self.width
+        else:
+            self.widget_diameter = self.height
```
Comment (agalbachicar) [inline src/rqt_gauges_2/steering_wheel_gauge.py]:
```suggestion
        self.widget_diameter = min(self.width, self.height)
```

File: src/rqt_gauges_2/steering_wheel_gauge.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        pen.setColor(QColor(self.text_color))
+        painter.setPen(pen)
+        painter.drawText(rect, Qt.AlignCenter, f"{self.value}{self.suffix}")
+
+        # End Painter
+        painter.end()
+
+        self.draw_big_scaled_marker(131, 278, 5)
+        self.create_scale_marker_values_text(131, 278, 5)
+
```
Comment (agalbachicar) [inline src/rqt_gauges_2/steering_wheel_gauge.py:150]: Same comment here.

File: src/rqt_gauges_2/steering_wheel_gauge.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+        if self.width <= self.height:
+            self.widget_diameter = self.width
+        else:
+            self.widget_diameter = self.height
+
+        self.resize(self.width, self.height)
+
+
+    def draw_big_scaled_marker(self, start_angle_value, angle_size, scalaCount):
```
Comment (agalbachicar) [inline src/rqt_gauges_2/steering_wheel_gauge.py]:
```suggestion
    def draw_big_scaled_marker(self, start_angle_value, angle_size, scala_count):
```

Comment (agalbachicar) [inline src/rqt_gauges_2/steering_wheel_gauge.py]: Can we set the types of the variables? I think that will help to strengthen the code. Here and there.

File: src/rqt_gauges_2/steering_wheel_gauge.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+        painter = QPainter(self)
+        painter.setRenderHint(QPainter.Antialiasing)
+        painter.translate(self.width / 2, self.height / 2)
+
+        self.pen = QPen(Qt.black)
```
Comment (agalbachicar) [inline src/rqt_gauges_2/steering_wheel_gauge.py]: Do we need to kep it in self?

File: src/rqt_gauges_2/steering_wheel_gauge.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+        self.pen.setWidth(2)
+        painter.setPen(self.pen)
+
+        painter.rotate(start_angle_value - self.angle_offset)
+        steps_size = (float(angle_size) / float(scalaCount))
+        scale_line_outer_start = (self.widget_diameter)/2
+        scale_line_lenght = (self.widget_diameter / 2) - (self.widget_diameter / 20)
```
Comment (agalbachicar) [inline src/rqt_gauges_2/steering_wheel_gauge.py]: Typo: scale_line_length

Comment (agalbachicar) [inline src/rqt_gauges_2/steering_wheel_gauge.py]: Can we pre-compute these values in the init?

File: src/rqt_gauges_2/steering_wheel_gauge.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+
+        pen_shadow = QPen()
+
+        pen_shadow.setBrush(QColor(0, 0, 0, 255))
+        painter.setPen(pen_shadow)
+
+        text_radius_factor = 0.75
+        text_radius = self.widget_diameter/2 * text_radius_factor
+
+        scale_per_div = int((self.maxValue - self.minValue) / scalaCount)
```
Comment (agalbachicar) [inline src/rqt_gauges_2/steering_wheel_gauge.py]: Can we precompute this at the init?

File: src/rqt_gauges_2/steering_wheel_gauge.py [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+            text = str(int(self.minValue + scale_per_div * i))
+            w = fm.width(text) + 1
+            h = fm.height()
+            painter.setFont(QFont(self.font_family, self.scale_font_size))
+            angle = angle_distance * i + float(scale_angle_start_value - self.angle_offset)
+            x = text_radius * math.cos(math.radians(angle))
+            y = text_radius * math.sin(math.radians(angle))
+
+            text = [x - int(w/2), y - int(h/2), int(w), int(h), Qt.AlignCenter, text]
+            painter.drawText(text[0], text[1], text[2], text[3], text[4], text[5])
```
Comment (agalbachicar) [inline src/rqt_gauges_2/steering_wheel_gauge.py]: Can we create a function and isolate a bit the functionality?

File: src/rqt_gauges_2/steering_wheel_gauge.py [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+        # Updates the value that the gauge is indicating.
+        # Args: 
+        #   value: Value to update the gauge with.
+  
+        if value <= self.minValue:
+            self.value = self.minValue
+        elif value >= self.maxValue:
+            self.value = self.maxValue
+        else:
+            self.value = value
```
Comment (agalbachicar) [inline src/rqt_gauges_2/steering_wheel_gauge.py]:
```suggestion
        value = max(value, self.minValue)
        value = min(value, self.maxValue)
```

File: src/rqt_gauges_2/steering_wheel_gauge.py [hunk 10]
```diff
@@ -0,0 +1,160 @@
+import math
+
+from PyQt5.QtWidgets import QWidget
+from PyQt5.QtGui import QPolygon, QPolygonF, QColor, QPen, QFont, QPainter, QFontMetrics, QConicalGradient, QRadialGradient, QBrush
+from PyQt5.QtCore import Qt, QPoint, QPointF, QObject, QRect
+
+class SteeringWheelGauge(QWidget):
```
Comment (agalbachicar) [inline src/rqt_gauges_2/steering_wheel_gauge.py:7]: Style preference: we have a lot of magic numbers everywhere. Can we use named constants?

**Review by agalbachicar — CHANGES_REQUESTED** [review]
Checkpoint and sorry for the late review.

**Comment by andrewbest-tri** [comment]
Tried to play around with this today, errors in setup.py (too many ] around the resource files) and launching the widgets causes errors for me in qt point. This machine has humble though, so maybe thats the problem

**Comment by andrewbest-tri** [comment]
steering wheel spawned but using it caused segfaults

**Comment by francocipollone** [comment]
@andrewbest-tri Sorry about that issue. Yes there was an error when the branch was updated.

I tried this locally in Foxy (as it Foxy was targeted).

After `colcon build & source install/setup.bash`
```
rqt --force-discover
```
On another terminal
```
ros2 topic pub /test_topic std_msgs/msg/Int16 "data: 10"

```


image

File: src/rqt_gauges_2/steering_wheel_gauge.py (hunk 8 above)
Comment (Voldivh) [reply src/rqt_gauges_2/steering_wheel_gauge.py]: Hmm we certainly can, however, it is using several local variables defined on the same method. Does it makes sense to create a function with all of those arguments that is only been used at that place in order to clean it up a little bit?

File: src/rqt_gauges_2/steering_wheel_gauge.py (hunk 9 above)
Comment (Voldivh) [reply src/rqt_gauges_2/steering_wheel_gauge.py]: Oh, I like this. Done.

File: src/rqt_gauges_2/steering_wheel_gauge.py (hunk 7 above)
Comment (Voldivh) [reply src/rqt_gauges_2/steering_wheel_gauge.py]: We can for `text_radius_factor` and `text_radius`. The variable `scale_per_div` depends on an argument from the method. Done.

File: src/rqt_gauges_2/steering_wheel_gauge.py (hunk 6 above)
Comment (Voldivh) [reply src/rqt_gauges_2/steering_wheel_gauge.py]: We can. Done.

Comment (Voldivh) [reply src/rqt_gauges_2/steering_wheel_gauge.py]: Thanks. Done.

File: src/rqt_gauges_2/steering_wheel_gauge.py (hunk 5 above)
Comment (Voldivh) [reply src/rqt_gauges_2/steering_wheel_gauge.py]: We don't, nice catch. Done.

File: src/rqt_gauges_2/steering_wheel_gauge.py (hunk 4 above)
Comment (Voldivh) [reply src/rqt_gauges_2/steering_wheel_gauge.py]: Sure and done.

Comment (Voldivh) [reply src/rqt_gauges_2/steering_wheel_gauge.py]: Done.

File: src/rqt_gauges_2/steering_wheel_gauge.py (hunk 2 above)
Comment (Voldivh) [reply src/rqt_gauges_2/steering_wheel_gauge.py]: Done.

File: src/rqt_gauges_2/steering_wheel.py (hunk 1 above)
Comment (Voldivh) [reply src/rqt_gauges_2/steering_wheel.py]: Done.

**Pull request merged**